Introduce new imfile module parameter: deleteStateOnFileMove

Cropi opened the pull request:
When set to on, this parameter ensures that the state file associated with a monitored file is deleted if the file is moved or rotated. By default, the state file is retained. Enabling this option prevents leftover state files from accumulating in the spool directory when monitored files are rotated and later deleted, which is important for many use cases, e.g. to not fill up spool directory with obsolete state files.

I attached a test case as well. If you turn off this new parameter, then the state file is not deleted at exit.

rgerhards commented:
@Cropi There are some real CI issue in the failures.

Also, I am a bit concerned about the non-backward compatibility. Can we reason why it is necessary or at least acceptable in the current case? In any case, doc needs to be updated. **Note: doc is now included in the main project under the `doc` subtree**.

Cropi commented on plugins/imfile/imfile.c:
```diff
@@ -2791,7 +2806,8 @@
 	getFullStateFileName(statefn, act->file_id, statefname, sizeof(statefname));
-	DBGPRINTF("persisting state for '%s', state file '%s'\n", act->name, statefname);
+	DBGPRINTF("persisting state for '%s', state file '%s', file-id: '%s', file_id_prev: '%s', ino: %lu\n",
+		act->name, statefname, act->file_id, act->file_id_prev, act->ino);
```
Fixed.

Cropi commented:
My use case is quite straightforward. I’m using the `imfile` module to monitor the `/var/log/input` file, and I have `logrotate` configured to rotate these logs once they reach a certain size. During rotation:

-  `/var/log/input.2` is renamed to `/var/log/input.3`
-  `/var/log/input.1` becomes `/var/log/input.2`
-  `/var/log/input` is rotated to `/var/log/input.1`
-  A new `/var/log/input` file is created

Rsyslog correctly continues monitoring the newly created file and stops tracking the rotated ones. However, the problem is that a separate state file remains on the system for each rotated file. As a result, I end up with four state files, even though I only need one. Over time, the number of these leftover state files keeps growing, even if the corresponding log files have been deleted.

I’m open to any changes or solutions that could help address this issue.

Cropi commented:
Hello,
Apologies, I may not have explained this fully or implemented it correctly before. I’ve now updated the code to introduce both module-level and input-level parameters for this feature. By default, these parameters are disabled, so they don’t affect existing behavior.

The goal of this PR is to enable removal of state files that are no longer needed, avoiding the accumulation of stale state files when monitored files are rotated or moved. Currently, when a monitored file is moved, its state file remains on the system. Over time, this leads to multiple unused “zombie” state files.

I’ve also added documentation for the new parameter, as well as a test case demonstrating that, by default, state files are not deleted. If you remove the `deleteStateOnFileMove` setting from the test, it will fail, showing the default behavior preserves state files.

> @Cropi Sry, I am a bit puzzled: the parameter currently exists, what is newly added is just the module parameter (which shall act as default). Another change is to the default of the equally-name pre-existing input parameter's default value. Plus some code changes. So the actual change semantically does not match the comit and PR message.
> 
> Does it instead:
> 
>     * add a new **module** parameter? (if so, I have not checked if it properply works as default-setting - the updated doc does not reflect that function)
> 
>     * a change the the default for state file deletion at the input() level
>       This is something that, after consideration, I think we shall not do. There is a sufficient number of installations which depend on the pre-existing default.
> 
>     * code changes that seem to point to and fix a bug in current code base?
> 
> 
> It would be great if you could review and/or clarify.
